### PR TITLE
[#70292] Primerize Custom Field forms

### PR DESCRIPTION
**Pull request by myabc**
# Ticket

https://community.openproject.org/wp/70292

# What are you trying to accomplish?
* Primerize the (not yet) custom field forms by extending the existing `DetailsComponent`.
* For Fields of type "List" there is a new tab "items" (like for hierarchy and weigthed item lists)
  * the tab shows the custom options 
  * the actual custom options table was untouched, because we plan to migrate the list fields into the hierarchy fields and can then use the existing components on the items tab 
  * The presence validation for list custom fields was removed

## Screenshots
Bildschirmfoto 2026-02-19 um 14 18 57

Bildschirmfoto 2026-02-19 um 14 18 42

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

**Review by myabc — COMMENTED** [review]
Since I'm not officially "here" this week (nor assigned), I'm not going to give this a full review. I did quickly want to check in and see if there were any issues with the foundation I created earlier in the year.

While I agree that it makes sense to have one way of rendering forms, I'm a bit disappointed to see everything moved into `DetailsForm`. I'm a big fan of the composability that the Primer Forms DSL offers. I also made it a goal to remove a lot of the conditional logic that was in `_form.html.erb`, so this feels like a step back. Indeed @ulferts had suggested that we create an analogue to `supports?`, but for custom field class, so for me that would have been a logical next step.

That said, I might have been wrong about letting individual forms define which field type/custom field class they support, but I do think this use case screams for some sort of registry-pattern like implementation (I know we were discussing this briefly, in the context of inline editing).

This PR had been on hold on my side for two reasons:

a) editing available options was only partially primerized and not working correctly, _and_
b) there were no tests despite the non-trivial logic involved.

It looks like (a) is now resolved (it makes sense to undo that work) — thank you for addressing that 👍
For (b), I’d still strongly recommend adding some component spec coverage before we merge. Given the amount of state handling here, having tests around the main flows and edge cases would significantly reduce risk of regressions.

**Comment by myabc** [comment]
My compromise suggestion would be to define a registry of custom field rules. For the time-being this could live at the top of the `DetailsForm` class.

```ruby
   # Registry defining field visibility rules.
   #
   # Each entry maps a field name to a set of constraints:
   #   formats:  { only: [...] } or { except: [...] } — matches against model.field_format
   #   models:   { only: [...] } — matches against model.is_a?(klass); omit to allow all model types
   #   if:       Symbol — an additional method name called on the model that must return true
   #
   FIELD_VISIBILITY = {
     section: {
       models: { only: [ProjectCustomField] }
     },
     default_bool: {
       formats: { only: %w[bool] }
     },
     default_text: {
       formats: { except: %w[list bool date text user version hierarchy weighted_item_list calculated_value] }
     },
     default_rich_text: {
       formats: { only: %w[text] }
     },
     min_max: {
       formats: { except: %w[list bool date user version link hierarchy weighted_item_list calculated_value] }
     },
     regex: {
       formats: { except: %w[list bool date user version hierarchy weighted_item_list calculated_value] }
     },
     formula: {
       formats: { only: %w[calculated_value] }
     },
     multi_value: {
       if: :multi_value_possible?
     },
     non_open_versions: {
       formats: { only: %w[version] },
       if: :allow_non_open_versions_possible?
     },
     is_required: {
       formats: { except: %w[calculated_value bool] }
     },
     is_for_all: {
       models: { only: [WorkPackageCustomField, ProjectCustomField] }
     },
     is_filter: {
       models: { only: [WorkPackageCustomField] }
     },
     is_searchable: {
       models: { only: [WorkPackageCustomField, ProjectCustomField] },
       formats: { except: %w[bool date float int user version hierarchy weighted_item_list calculated_value] }
     },
     right_to_left: {
       models: { only: [WorkPackageCustomField, ProjectCustomField] },
       formats: { only: %w[text] }
     },
     admin_only: {
       models: { only: [ProjectCustomField, UserCustomField] }
     },
     editable: {
       models: { only: [UserCustomField] }
     }
   }.freeze

   def show_field?(field_name)
     rules = FIELD_VISIBILITY.fetch(field_name)
     if (model_rule = rules[:models])
       return false unless model_rule[:only].any? { |klass| model.is_a?(klass) }
     end

      if (format_rule = rules[:formats])
        if format_rule.key?(:only)
          return false unless format_rule[:only].include?(model.field_format)
        elsif format_rule.key?(:except)
          return false if format_rule[:except].include?(model.field_format)
        end
      end

      if (condition = rules[:if])
        return false unless model.public_send(condition)
      end

      true
    end
```

There would then be one guard `show_field?` method, e.g.:

```ruby
    if show_field?(:min_max)
        details_form.text_field(
          name: :min_length,
          type: :number,
          label: label(:min_length),
          caption: instructions(:min_max),
          input_width: :small
        )

        details_form.text_field(
          name: :max_length,
          type: :number,
          label: label(:max_length),
          caption: instructions(:min_max),
          input_width: :small
        )
      end

      if show_field?(:regex)
        details_form.text_field(
          name: :regexp,
          label: label(:regexp),
          caption: instructions(:regexp),
          input_width: :medium
        )
      end
```

File: app/forms/custom_fields/details_form.rb [hunk 1]
```diff
@@ -60,11 +60,31 @@ class DetailsForm < ApplicationForm
...
-      if show_multi_value_field?
-        details_form.check_box(
-          name: :multi_value,
-          label: label(:multi_value),
-          caption: instructions(:multi_select)
+      if show_min_max_field?
+        details_form.text_field(
+          name: :min_length,
+          type: :number,
+          label: label(:min_length),
```
Comment (myabc) [inline app/forms/custom_fields/details_form.rb:67]: we should use `attribute_name` not this custom `label` method. `attribute_name` (and in turn, `Model.human_attribute_name`) is inheritance-aware when looking up translations. This is from memory, but the cascade looks something like:

```
activerecord.attributes.project_custom_field.min_length
activerecord.attributes.custom_field.min_length
activerecord.attributes.min_length
attributes.project_custom_field.min_length
attributes.custom_field.min_length
attributes.min_length
```

File: app/forms/custom_fields/details_form.rb [hunk 2]
```diff
@@ -60,11 +60,31 @@ class DetailsForm < ApplicationForm
...
+          caption: instructions(:min_max),
+          input_width: :small
+        )
+      end
+
+      if show_regex_field?
+        details_form.text_field(
+          name: :regexp,
+          label: label(:regexp),
+          size: 50,
```
Comment (myabc) [inline app/forms/custom_fields/details_form.rb]: I don't think Primer supports a `size` option.

File: config/locales/en.yml [hunk 3]
```diff
@@ -417,6 +417,12 @@ en:
           Note that only with 'For all projects' selected, the custom field will show up in global views.
       formula:
         project: "Add numeric values or type / to search for an attribute or a mathematical operator."
+      regexp:
+        all: "eg. ^[A-Z0-9]+$"
+        project: "eg. ^[A-Z0-9]+$"
+      min_max:
+        all: "0 means no restriction"
+        project: "0 means no restriction"
```
Comment (myabc) [inline config/locales/en.yml:438]: The `instructions` method could also be reworked to be slightly more intelligent / use I18n fallbacks, making this repetition unnecessary:

```ruby
def instructions(field)
  model_scope = model.class.name.demodulize.underscore
  I18n.t(
    "custom_fields.instructions.#{field}.#{model_scope}",
    default: I18n.t("custom_fields.instructions.#{field}.all")
  )
end
```

**Comment by HDinger** [comment]
I knew that you would not like it. However, the solution was not as progressed as we thought last week. It was lacking support for model checks, certain fields were missing, the field checks were partially wrong and what was the most important reason for me to switch: It was introducing a new way of rendering the form without migrating the already existing code. I had to make decision in which direction to go and decided to use the one we already had. We discussed today on how to continue here and decided to stick with the current approach for now and migrate everything all together in a separate next step. 

Regarding the lack of test coverage, I can't really follow. There might not be a component spec but the form is nevertheless heavily tested. There are mutliple tests assuring the correct fields are shown as well as the creation flow especially for lists.

File: app/components/table_component.rb [hunk 4]
```diff
@@ -31,8 +31,17 @@
...
     super(rows, **)
+
+    @system_arguments = table_arguments
+    @system_arguments[:tag] = :table
+    @system_arguments[:classes] = class_names(
+      @system_arguments[:classes],
+      "generic-table"
+    )
+    @system_arguments[:data] ||= {}
+    @system_arguments[:data][:controller] = "table-highlighting"
```
Comment (bsatarnejad) [inline app/components/table_component.rb]: @system_arguments[:data][:controller] = "table-highlighting" will overwrite any data-controller passed in table_arguments.
Could we append instead (e.g., merge existing controllers + "table-highlighting") so callers don’t lose their controller?

File: app/controllers/concerns/custom_fields/shared_actions.rb [hunk 5]
```diff
@@ -139,14 +157,18 @@ def delete_custom_values!(custom_option)
...
         index = 0
 
-        params[:custom_field][:custom_options_attributes].each_value do |attributes|
+        custom_options_attributes.each_value do |attributes|
           attributes[:position] = (index = index + 1)
         end
       end
+
+      def custom_options_attributes
+        params[:custom_field][:custom_options_attributes]
```
Comment (bsatarnejad) [inline app/controllers/concerns/custom_fields/shared_actions.rb:172]: It is better to guard params[:custom_field].present? to avoid raising any error when there is no params[:custom_field]

**Review by bsatarnejad — CHANGES_REQUESTED** [review]
Image
Image
Image
Image

As you can see in the images above, the projects tab is expanded to the full viewport width, while the others are not. It is an inconsistency.

**Comment by HDinger** [comment]
> As you can see in the images above, the projects tab is expanded to the full viewport width, while the others are not. It is an inconsistency.

This is existing (inconsistent) behaviour and was not changed in the scope of this PR. So I'd prefer to keep it out of scope given the size of this PR.
